Modules/blake2module.c

```diff
@@ -0,0 +1,925 @@
+        goto error;
+    }
+
+    /* Validate salt parameter. */
+    if ((salt->obj != NULL) && salt->len) {
+        if (salt->len > (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_SALT_BYTES : HACL_HASH_BLAKE2S_SALT_BYTES)) {
+            PyErr_Format(PyExc_ValueError,
+                "maximum salt length is %d bytes",
+                (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_SALT_BYTES : HACL_HASH_BLAKE2S_SALT_BYTES));
+            goto error;
+        }
+        memcpy(salt_, salt->buf, salt->len);
+    }
+
+    /* Validate personalization parameter. */
+    if ((person->obj != NULL) && person->len) {
+        if (person->len > (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_PERSONAL_BYTES : HACL_HASH_BLAKE2S_PERSONAL_BYTES)) {
+            PyErr_Format(PyExc_ValueError,
+                "maximum person length is %d bytes",
+                (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_PERSONAL_BYTES : HACL_HASH_BLAKE2S_PERSONAL_BYTES));
+            goto error;
+        }
+        memcpy(personal_, person->buf, person->len);
+    }
+
+    /* Validate tree parameters. */
+    if (fanout < 0 || fanout > 255) {
+        PyErr_SetString(PyExc_ValueError,
+                "fanout must be between 0 and 255");
+        goto error;
+    }
+
+    if (depth <= 0 || depth > 255) {
+        PyErr_SetString(PyExc_ValueError,
+                "depth must be between 1 and 255");
+        goto error;
+    }
+
+    if (leaf_size > 0xFFFFFFFFU) {
+        PyErr_SetString(PyExc_OverflowError, "leaf_size is too large");
+        goto error;
+    }
+
+    if (is_blake2s(self->impl) && node_offset > 0xFFFFFFFFFFFFULL) {
+        /* maximum 2**48 - 1 */
+         PyErr_SetString(PyExc_OverflowError, "node_offset is too large");
+         goto error;
+     }
+
+    if (node_depth < 0 || node_depth > 255) {
+        PyErr_SetString(PyExc_ValueError,
+                "node_depth must be between 0 and 255");
+        goto error;
+    }
+
+    if (inner_size < 0 ||
+        (unsigned) inner_size > (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_OUT_BYTES : HACL_HASH_BLAKE2S_OUT_BYTES)) {
+        PyErr_Format(PyExc_ValueError,
+                "inner_size must be between 0 and is %d",
+                (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_OUT_BYTES : HACL_HASH_BLAKE2S_OUT_BYTES));
+        goto error;
+    }
+
+    /* Set key length. */
+    if ((key->obj != NULL) && key->len) {
+        if (key->len > (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_KEY_BYTES : HACL_HASH_BLAKE2S_KEY_BYTES)) {
+            PyErr_Format(PyExc_ValueError,
+                "maximum key length is %d bytes",
+                (is_blake2b(self->impl) ? HACL_HASH_BLAKE2B_KEY_BYTES : HACL_HASH_BLAKE2S_KEY_BYTES));
+            goto error;
+        }
+    }
+
+    // Unlike the state types, the parameters share a single (client-friendly)
+    // structure.
+
+    Hacl_Hash_Blake2b_blake2_params params = {
+        .digest_length = digest_size,
+        .key_length = key->len,
```

Comment:
There is a warning here due to a conversion of `Py_ssize_t` to `uint8_t`. Whlie it shouldn't be the import at runtime, since have the `key->len > ...` check above, please add an explicit cast.